File: macos/BinderClipMac/Sources/Direct/SessionLiveness.swift
import Foundation

enum SessionLiveness {
    static let heartbeatBudget: TimeInterval = 5

    static func ipv4Bytes(_ ip: String) -> [UInt8]? {
        let parts = ip.split(separator: ".")
        guard parts.count == 4 else { return nil }
        let bytes = parts.compactMap { UInt8($0) }
        guard bytes.count == 4 else { return nil }
        return bytes
    }

    static func matchingPrefixBits(remote: String, local: String) -> Int {
        guard let remoteBytes = ipv4Bytes(remote), let localBytes = ipv4Bytes(local) else { return -1 }
        var bits = 0
        for index in 0..<4 {
            if remoteBytes[index] == localBytes[index] {
                bits += 8
                continue
            }
            let xor = remoteBytes[index] ^ localBytes[index]
            for shift in stride(from: 7, through: 0, by: -1) {
                if xor & (1 << shift) != 0 { return bits }
                bits += 1
            }
        }
        return bits
    }

    /// Longest IPv4 prefix match. Mesh remotes (`100.`) prefer a local `100.` address.
    static func boundLocalAddress(remote: String, localAddresses: [String]) -> String? {
        let locals = localAddresses.filter { ipv4Bytes($0) != nil }
        guard !locals.isEmpty else { return nil }
        if remote.hasPrefix("100."), let mesh = locals.first(where: { $0.hasPrefix("100.") }) {
            return mesh
        }
        return locals.max { lhs, rhs in
            matchingPrefixBits(remote: remote, local: lhs) < matchingPrefixBits(remote: remote, local: rhs)
        }
    }

    static func shouldEvict(boundLocal: String?, currentLocals: [String]) -> Bool {
        guard let boundLocal, ipv4Bytes(boundLocal) != nil else { return false }
        return !currentLocals.contains(boundLocal)
    }

    static func isAlive(
        boundLocal: String?,
        currentLocals: [String],
        lastHeard: Date?,
        now: Date,
        budget: TimeInterval = heartbeatBudget
    ) -> Bool {
        guard let lastHeard else { return false }
        if shouldEvict(boundLocal: boundLocal, currentLocals: currentLocals) { return false }
        return now.timeIntervalSince(lastHeard) <= budget
    }
}

File: android/app/src/main/java/net/wastu/binderclip/SessionLiveness.kt
package net.wastu.binderclip

object SessionLiveness {
    fun ipv4Bytes(ip: String): ByteArray? {
        val parts = ip.split('.')
        if (parts.size != 4) return null
        val bytes = ByteArray(4)
        for (index in 0..3) {
            val value = parts[index].toIntOrNull() ?: return null
            if (value !in 0..255) return null
            bytes[index] = value.toByte()
        }
        return bytes
    }

    fun matchingPrefixBits(remote: String, local: String): Int {
        val remoteBytes = ipv4Bytes(remote) ?: return -1
        val localBytes = ipv4Bytes(local) ?: return -1
        var bits = 0
        for (index in 0 until 4) {
            val remoteOctet = remoteBytes[index].toInt() and 0xff
            val localOctet = localBytes[index].toInt() and 0xff
            if (remoteOctet == localOctet) {
                bits += 8
                continue
            }
            val xor = remoteOctet xor localOctet
            for (shift in 7 downTo 0) {
                if (xor and (1 shl shift) != 0) return bits
                bits += 1
            }
        }
        return bits
    }

    fun boundLocalAddress(remote: String, localAddresses: List<String>): String? {
        val locals = localAddresses.filter { ipv4Bytes(it) != null }
        if (locals.isEmpty()) return null
        if (remote.startsWith("100.")) {
            locals.firstOrNull { it.startsWith("100.") }?.let { return it }
        }
        return locals.maxByOrNull { matchingPrefixBits(remote, it) }
    }

    fun shouldEvict(boundLocal: String?, currentLocals: List<String>): Boolean {
        val bound = boundLocal ?: return false
        if (ipv4Bytes(bound) == null) return false
        return bound !in currentLocals
    }

    fun isAlive(
        boundLocal: String?,
        currentLocals: List<String>,
        lastHeardMs: Long?,
        nowMs: Long,
        budgetMs: Long = SyncProtocol.HEARTBEAT_BUDGET_MS,
    ): Boolean {
        if (lastHeardMs == null) return false
        if (shouldEvict(boundLocal, currentLocals)) return false
        return nowMs - lastHeardMs <= budgetMs
    }
}
